server/data/data-jmap-postgres/src/main/java/org/apache/james/jmap/postgres/identity/PostgresCustomIdentityDAO.java

```diff
@@ -0,0 +1,233 @@
+import com.github.fge.lambdas.Throwing;
+import com.google.common.collect.ImmutableList;
+
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+import reactor.core.scala.publisher.SMono;
+import scala.Option;
+import scala.collection.JavaConverters;
+import scala.collection.immutable.List;
+import scala.collection.immutable.Seq;
+import scala.jdk.javaapi.OptionConverters;
+import scala.runtime.BoxedUnit;
+
+public class PostgresCustomIdentityDAO implements CustomIdentityDAO {
+    static class Email {
+        private String name;
+        private String email;
+
+        public Email() {
+        }
+
+        public Email(String name, String email) {
+            this.name = name;
+            this.email = email;
+        }
+
+        public String getName() {
+            return name;
+        }
+
+        public void setName(String name) {
+            this.name = name;
+        }
+
+        public String getEmail() {
+            return email;
+        }
+
+        public void setEmail(String email) {
+            this.email = email;
+        }
+    }
+
+    private final PostgresExecutor.Factory executorFactory;
+    private final ObjectMapper objectMapper = new ObjectMapper();
+
+    @Inject
+    public PostgresCustomIdentityDAO(PostgresExecutor.Factory executorFactory) {
+        this.executorFactory = executorFactory;
+    }
+
+    @Override
+    public Publisher<Identity> save(Username user, IdentityCreationRequest creationRequest) {
+        return save(user, IdentityId.generate(), creationRequest);
+    }
+
+    @Override
+    public Publisher<Identity> save(Username user, IdentityId identityId, IdentityCreationRequest creationRequest) {
+        final Identity identity = creationRequest.asIdentity(identityId);
+        return upsertReturnPublisher(user, identity);
+    }
+
+    @Override
+    public Publisher<Identity> list(Username user) {
+        return executorFactory.create(user.getDomainPart())
+            .executeRows(dslContext -> Flux.from(dslContext.selectFrom(TABLE_NAME)
+                .where(USERNAME.eq(user.asString()))))
+            .map(Throwing.function(this::readRecord));
+    }
+
+    @Override
+    public SMono<Identity> findByIdentityId(Username user, IdentityId identityId) {
+        return SMono.fromPublisher(findByIdentityIdReturnPublisher(user, identityId));
+    }
+
+    private Publisher<Identity> findByIdentityIdReturnPublisher(Username user, IdentityId identityId) {
+        return executorFactory.create(user.getDomainPart())
+            .executeRow(dslContext -> Mono.from(dslContext.selectFrom(TABLE_NAME)
+                .where(USERNAME.eq(user.asString()))
+                .and(ID.eq(identityId.id()))))
+            .map(Throwing.function(this::readRecord));
+    }
+
+    @Override
+    public Publisher<BoxedUnit> update(Username user, IdentityId identityId, IdentityUpdate identityUpdate) {
+        return Mono.from(findByIdentityIdReturnPublisher(user, identityId))
```

Comment:
Do you think we should refactor here,
We don't need `find` and `update`, we can `update` directly in a single request?
I checked the `IdentityUpdate` class, it is quite simple, no dep business.
WDYT?

Reply:
It require refactoring Identity which is quite complicated IMO. I think let it be for now